Task 14

gusevpa opened the pull request:
Рабочий вариант парсера тинькофского API, возвращающий двумя методами(покупка/продажа) свежий курс доллара.

gusevpa committed:
Класс/метод, начало работы с тинькофф по курсам валют

AlexeyKorban committed:
Task-17

Add dictionary interfaces

gusevpa committed:
Add TinkoffParserTest.java

gusevpa committed:
Update TinkoffParser
Add timer

AlexeyKorban committed:
Task-20

Add CustomerRepository interface

AlexeyKorban committed:
Task-22

Fix readme, add program arguments

NameNickNo committed:
Nickolay: Creating CustomerRepisitoryImpl, modified sql-query.properties

NameNickNo committed:
Modified person_schema.sql and sql-query.properties in test

NameNickNo committed:
Nickolay: Creating 2 scripts for testing, customer_data and customer_schema

NameNickNo committed:
Nickolay: replace phone -> phonenumber int scripts, creating tests -> CustomerTestData, CustomerRepositoryImplTest, modified -> customer_data, customer_schema. Issues are constructors and toString()

NameNickNo committed:
Nickolay: correcting CustomerRepositoryImpl, optional.ofNullable

NameNickNo committed:
Nickolay: Modified CustomerRepositoryImpl, add new tests on nulls CustomerRepositoryImplTest

gusevpa committed:
CustomerEntity пустой конструктор

NameNickNo committed:
Nickolay: Create DictionaryRepositoryImpl

NameNickNo committed:
Nickolay: Add two sql scrypts, add StatusDto, add StatusEntity, add test on empty list

NameNickNo committed:
Nickolay: modified sql-query.properties

NameNickNo committed:
Nickolay: create method getAllStatuses, test is success, modified scripts

NameNickNo committed:
Nickolay: create DictionaryControllerImpl and test for it, modify repository and controller intefaces, add method in StatusTestData

NameNickNo committed:
Nickolay: change type of code in Status, add test in DictionaryControllerImplTest, add method in StatusTestData

NameNickNo committed:
Nickolay: create second method in DictionaryControllerImpl

NameNickNo committed:
Nickolay: fix unuses imports

NameNickNo committed:
Nickolay: fix scrypts, rename DictionaryController to DictionaryService

NameNickNo committed:
Nickolay: Add the ability to find by tag, add 2 tests in CustomerRepositoryTest

NameNickNo committed:
Nickolay: add 3 tests in CustomerRepositoryTest, add 2 methods in CustomerTestData. Task is complete

NameNickNo committed:
Nickolay: remove unused import

AlexeyKorban committed:
Task-25

Add codacy badge

Commit:
Task-26 Change DataSource to PostgreSQL

Commit:
Task-26 bugfix

NameNickNo committed:
Nickolay: find bug, equals and hashcode ind StatusEntity

yuliaPlatonova committed:
Task28(Dictionary Controller added)

yuliaPlatonova committed:
Task28(DictionaryControllerTest added)

yuliaPlatonova committed:
Task29(SQL scripts for dictionary table added)

yuliaPlatonova committed:
Task29(refactoring of some tests)

yuliaPlatonova committed:
Task29(small refactoring)

gusevpa committed:
Дописал рабочие геттеры.
Возвращают BigDecimal курс доллара покупка/продажа.

gusevpa committed:
Merge branch 'main' into Task-14

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+public class TinkoffParser {
+
+    private static LocalDateTime usdLastResp; //возраст курса доллара
```
Я бы переименовал. Например, timeLastUpdate.
Ведь мы там храним время, а не usd. 
А как перевести Resp - я вообще не понял)))

gusevpa replied:
Переименовал в lastUsdCostUpdateTime. Теперь сразу ясно что это за зверь)

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+    private static LocalDateTime usdLastResp; //возраст курса доллара
+    private static BigDecimal usdToRub; //курс доллара
+    private static BigDecimal rubToUsd; //обратный курс
```
Я бы переименовал. Ведь по сути нам нужна цена покупки и продажи доллара.
Например, costBuyDollar и costSellDollar. Типо цена покупки 77 рублей, продажи 74 рубля.
Предполагается, что продаём - мы банку. У банка названия курсов могут быть зеркальными.
Обратный курс - нам вообще не нужен, его всегда можно вывести из основного.

gusevpa replied:
costUsdBuy и costUsdSell в новой версии. Никаких туда-сюда)

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+    private static BigDecimal usdToRub; //курс доллара
+    private static BigDecimal rubToUsd; //обратный курс
+
+    public static void init(){
```
Я бы сделал его приватным, и добавил принудительный его вызов, если время последнего обновления равно нулю. Ну чтобы не получать NPE, когда забыли сначала вызвать init.

gusevpa replied:
Добавил в пустой конструктор. Теперь само при создании обновляется.

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+        return Math.abs(p.getSeconds()) > 30L; //если разница p превышает 30 секунд, то возвращаем false
+    }
+
+    private static void usdReNewer () throws Exception{ //обновлятель курса доллара и его возраста
```
На мой взгляд - очень много мелких методов. Я бы объединил методы init, timeChecker и usdReNewer в один, цель которого - обновить курс. Назвал бы updateCostDollar. В нём бы проверял - что время последнего обновления - не нулевое, что время прошло больше 30 секунд, и только потом обновлял бы курс.

При этом вызов этого метода - принудительно бы запихнул в во все публичные методы.

AlexeyKorban replied:
Тут вопрос для обсуждения. Мне кажется лучше мелкие методы, с понятными названиями, чтобы сразу видеть что он делает. А не один большой метод со сложной логикой. Если все объединить то он у нас и инициализировать должен и время проверять и данные обновлять. Не очень жирно будет, но тенденция не хорошая

gusevpa replied:
Мелкие методы оставил. Позже коль появится желание вернуться - еще чуть раздроблю тот большой. Потому что так тесты писать проще будет.

jads84 commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+        InputStream is = new URL(url).openStream();
+        try {
+            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
+            String jsonText = readAll(rd);
```
Заменил вот так и всё работает:
```
            //String jsonText = readAll(rd);
            String jsonText = rd.readLine();;

```
Метод readAll - можно выкидывать.

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+    //Метод, запрашивающий api.tinkoff.ru, получающий и парсящий ответный JSON на предмет актуального курса доллара
+
+    private static JSONObject readJsonFromUrl(String url) throws IOException, JSONException {
```
Метод можно удалить. Смотри tinkRespUsdRub();

gusevpa replied:
Убрал.

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    private static String readAll(Reader rd) throws IOException {
```
Метод можно удалить. Смотри tinkRespUsdRub();

AlexeyKorban replied:
Согласен, тут что-то не так. Мы создали BufferedReader и теперь читаем по символам

gusevpa replied:
Согласен. Теперь читаем все символы сразу.

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+    private static void tinkRespUsdToRub() throws Exception {
+
+        JSONObject json = readJsonFromUrl("https://api.tinkoff.ru/v1/currency_rates");
```
Можно сократить вот так:

//JSONObject json = readJsonFromUrl("https://api.tinkoff.ru/v1/currency_rates");
String url = "https://api.tinkoff.ru/v1/currency_rates";
byte[] chars = new URL(url).openStream().readAllBytes();
JSONObject json = new JSONObject(new String(chars, "UTF-8"));

gusevpa replied:
Епонский бох.
Оно работает!

gusevpa replied:
Запилил!

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+public class TinkoffParser {
+
+    private static LocalDateTime usdLastResp; //возраст курса доллара
```
Предлагаю все переменные и методы сделать не статическими. Будем парсер хранить в контексте спринга и получать там где он необходим.

gusevpa replied:
Если спринг сам обернет это в синглтон - окей, без вопросов. Сделал.
Надеюсь я в скором времени смогу разобраться в этой магии)

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+    private static BigDecimal usdToRub; //курс доллара
+    private static BigDecimal rubToUsd; //обратный курс
+
+    public static void init(){
```
Как вариант можно время инициализировать в конструкторе, а данные обновлять после первого запроса. Зачем нам их сразу получать, если они могут и не понадобиться?

AlexeyKorban replied:
По хорошему стоит подумать как вообще убрать этот init

gusevpa replied:
Инит засунут в конструктор. Что делать дальше - подумаем потом.

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    private static boolean timeChecker (){
```
Результат этой проверки всегда инвертируется, может имеет смысл сразу проверять < 30L?

gusevpa replied:
Нене, переделал все остальное чтоб оставить проверку в таком виде.
Ну оно просто изначально криво было.
Пофикшено.

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+            return rubToUsd;
+        } else {
+            usdReNewer();
+            return rubToUsd;
```
В таких местах (где есть повторение кода) имеет смысл выносить его отдельно. Т.е. return вынести после if

gusevpa replied:
Done.

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+    private static JSONObject readJsonFromUrl(String url) throws IOException, JSONException {
+        InputStream is = new URL(url).openStream();
+        try {
+            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
```
Тут лучше использовать стандартные кодировки StandardCharsets.UTF_8. Надежней

gusevpa replied:
Окей, учтено.

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,118 @@
+
+    private static JSONObject readJsonFromUrl(String url) throws IOException, JSONException {
+        InputStream is = new URL(url).openStream();
+        try {
```
Лучше посмотреть try с ресурсами

gusevpa replied:
Согласен, пропустил. Больше не буду слепо копировать со stackowerflow и не проверять если заработало)

AlexeyKorban commented on src/test/java/ru/ldwx/crm/TinkoffParserTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class TinkoffParserTest {
```
Где тесты?)

gusevpa replied:
какие тесты?)))

gusevpa replied:
Нене, нет тестов)

gusevpa committed:
Update TinkoffParser.java

gusevpa committed:
Update TinkoffParser.java

Поправил все замечания. Все еще сомневаюсь насчет удаления вообще всех статиков, поэтому добавил всякую бяку типа синхронизации. Понимаю что криво, но почему - объяснить не могу. Надеюсь что мне расскажут)
При тестировании работоспособности обнаружил что иногда появляется null на выходе.

gusevpa commented:
Пофиксил. Сломалось в другом месте.
При использовании иногда новосозданный объект возвращает null вместо курса доллара (см. пример вывода).
rubToUSD - опрос заранее созданного объекта - всегда корректно
NEW rubToUSD - опрос объекта сразу при создании - иногда null

74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
null NEW rubToUSD
null NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB
74.05 rubToUSD
77.85 usdToRUB
74.05 NEW rubToUSD
77.85 NEW usdToRUB

gusevpa commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    public BigDecimal getCostUsdSell() throws Exception{ //возвращаем актуальный курс
+
```
полагаю что появление null на выходе как-то связано с тем, что некорректно описаны эти два public метода и их взаимодействие с usdRenewer(). Скорее всего надо сделать один метод, который возвращает массив из двух элементов BigDecimal и boolean значения, говорящего о наличии проблем со свежим курсом (интернет кончился, тинькоф упал etc).

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,101 @@
+        LocalDateTime nowDateTime = LocalDateTime.now(); //получаем текущее время системы nowDateTime
+        Duration p = Duration.between(lastUsdCostUpdateTime, nowDateTime); //получаем разницу между последним обновлением курса usd и текущей nowDateTime
+
+        return Math.abs(p.getSeconds()) > 30L; //если разница p превышает 30 секунд, то возвращаем false
```
Math.abs необходим чтобы убедиться что время не идет назад?)

gusevpa replied:
Именно! А то мало ли что!
Убрал.

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,101 @@
+
+public class TinkoffParser {
+
+    private static LocalDateTime lastUsdCostUpdateTime; //возраст курса доллара
```
Это поле нам не нужно чтобы было статичным. Используем только внутри класса, создаем вместе с классом.

gusevpa replied:
Убрал

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,101 @@
+        lastUsdCostUpdateTime = LocalDateTime.now();
+        try {
+            getTinkoffCurrencyExchangeRates();
+        } catch (Exception e) {
```
Если что-то пойдет не так, то мы проглотим исключение и будем возвращать null как минимум 30 секунд.
Может дергать init при первом вызове метода получения курса? Нужно подумать как лучше будет. В идеале не нужно ходить за курсом пока сервис не дернут. Возможно курс вообще не понадобится в конкретной сессии, зачем тратить на это время и ресурсы?

gusevpa replied:
Вот это да. надо пофиксить. Но как - пока не представляю.
Потому что раз мы создали класс - надо ему задать курс.
Если он не понадобится в течении таймера обновления (30 секунд) - то один запрос погоды не сделает.
А проверки на нулевой курс - нету..
Недоделка. Как и нескольких других необходимых проверок. Но пока работает. Запилим позжее.

gusevpa replied:
Окей, осознал. Сам осознал - прошу заметить!
Теперь это выглядит так:
 try {
            getTinkoffCurrencyExchangeRates();
            lastUsdCostUpdateTime = LocalDateTime.now();
        } catch (Exception e) {
            e.printStackTrace();
        }

AlexeyKorban commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,101 @@
+        return costUsdSell;
+    }
+
+    private static synchronized boolean timeChecker (){
```
Тут синхронизация не нужна. Это приватный метод и используется только внутри класса. static соответственно тоже будет не нужен, т.к. поле лучше сделать внутри класса.

gusevpa replied:
Согласен, перегнул.

gusevpa committed:
Update TinkoffParser.java

gusevpa committed:
Update TinkoffParser.java

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,105 @@
+        init();
+    }
+
+    private void init(){
```
зачем init, если теперь есть конструктор?

gusevpa replied:
Точно. Ну ладо, пока так пусть будет.
Переделаем потом.

Dmitrii-Iakovenko commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,105 @@
+        byte[] chars;
+        try(InputStream inputStream = new URL(url).openStream()){
+            chars = inputStream.readAllBytes();
+        }
```
а что будет, если выскочит исключение и массив байтов не заполнится?

gusevpa replied:
А вот тут будет косяк. Что делать - надо подумать.
Есть предложения?
Теоретически исключение может выпасть только в случае если юрл недоступен, а это надо организовывать проверку связи. Каким образом? Идей много, но все кривые.

NameNickNo commented on src/main/java/ru/ldwx/crm/TinkoffParser.java:
```diff
@@ -0,0 +1,105 @@
+        for (int i = 0; i < arr.length(); i++){
+            JSONObject a = arr.getJSONObject(i);
+
+            if(a.get("category").equals("DebitCardsTransfers") & a.getJSONObject("toCurrency").get("name").equals("RUB") & a.getJSONObject("fromCurrency").get("name").equals("USD")){
```
тяжело читается эта строка, плюс длина, неплохо бы разделить на две, тогда уж

gusevpa commented:
Йо-хо-хо!
Я переделаю. Обязательно.
Но пока работает.
Спасибо за замечания! Те, которые я не исправил в данном PR , будут учтены в последующих.
